README.md
# WHMCS - Módulo de PIX e Boleto PagHiper

Compatível com PHP 7.4, PHP 8.1 e WHMCS 8.

* Seu boleto sai direto no PDF anexo a fatura (funcionalidade exclusiva).
* Boletos registrados conforme especificação da FEBRABAN. Baixa de pagamentos automática.
* Emissão de PIX de forma simplificada (com retorno automático).
* Emita boletos bancários direto do seu WHMCS.
##
* [Funcionalidades](#funcionalidades)
* [Como instalar](#instalação)
* [Como configurar](#configuração)
* [Solicitação de suporte e novas funcionalidades](#solicitação-de-suporte-e-novas-funcionalidades)

## Funcionalidades
### Funcionalidades gerais para boleto e PIX
* **Taxa percentual**: porcertagem aplicada ao valor da fatura por usar o PagHiper.
* **Taxa fixa**: valor, em reais, somado ao valor da fatura por usar o PagHiper.
* **Exibe ou não a frase fixa**: configurada no painel, no site do PagHiper
* **Valor máximo para pagamento**: valor máximo para realizar pagamento por boleto/PIX.
* **Valor mínimo para pagamento**: valor mínimo para realizar pagamento por boleto/PIX.
* **Isentar de taxas**: não aplicar taxas (Taxa percentual + Taxa fixa) caso o valor da fatura seja maior ou igual ao definido.
### Funcionalidades para PIX
* **Desconto por pagto. realizado via PIX**: percentual retirado do valor da fatura, caso o pagamento ocorra por PIX.
#### Desconto por critério
* **Critério do desconto**: quando aplicar o desconto acima? Atualmente, o módulo suporta o critério "Apenas para novos serviços".
* **Porcentagem do desconto por critério**: percentual retirado do valor da fatura caso o critério acima se for atendido.
### Funcionalidades para boleto
* **Abrir boleto ao abrir fatura**: gera o boleto e o abre automaticamente quando a página de fatura é acessada.
* **Tolerância para pagto**: número máximo de dias em que o boleto poderá ser pago após o vencimento.
* **Vencimento padrão para boletos emitidos**: escolha a quantidade de dias para o vencimento de boletos reemitidos (para faturas ja vencidas).
* **Percentual da multa por atraso (%)**: o percentual máximo autorizado é de 2%, de acordo artigo 52, parágrafo primeiro do Código de Defesa do Consumidor, Lei 8.078/90.
* **Juros proporcional**: aplica 1% de juros máximo ao mês, esse percentual será cobrado proporcionalmente aos dias de atraso. Dividindo 1% por 30 dias = 0,033% por dia de atraso.
#### **Desconto por pagamento antecipado**
* **Qtde. de dias para aplicação de desconto**: número de dias em que o pagamento pode ser realizado com antecedência recebendo o desconto extra.
* **Desconto por pagto. antecipado**: percentual do desconto que será aplicado caso o pagamento ocorra de forma antecipada.
* **Gerar boletos para todos os pedidos?**: automaticamente gera boletos para todos os pedidos criados, independentemente do gateway de pagamento selecionado.
* **Validar campos de CPF/CNPJ no checkout?**: fornece opções para validar ou o CPF e o CNPJ, ou apenas um dos dois, ou nenhum.

## Funcionalidades avançadas

### Anexar PDF do boleto/PIX ao PDF da fatura
Copie e cole o código abaixo no início do arquivo `invoicepdf.tpl` do seu tema.

```php
<?php include dirname(__FILE__).'/../../modules/gateways/paghiper/inc/helpers/attach_pdf_slip.php'; ?>
```

### inserir código PIX e linha digitável de boletos a e-mails

Para essa funcionalidade, edite seu template de e-mail `no menu superior, vá com o mouse sobre a chave de fenda > clique em Opções > Pesquise por "modelos > Procure e abra o modelo que deseja implementar a funcionalidade"`.

Você pode usar dois campos de mesclagens, um para boleto e outro para PIX, respectivamente: `{$linha_digitavel}` e `{$codigo_pix}`. Basta inserí-los nos templates de comunicação de e-mail nos locais desejados.

## Como instalar e configurar
### Instalação

1. Crie sua conta na PagHiper: [clique aqui para saber como criar](https://www.paghiper.com/duvidas/como-se-cadastrar-no-paghiper/).

2. Baixe a última versão do gateway, [clicando aqui](https://github.com/LinkNacional/whmcs-boleto-pix/releases/latest/download/whmcs-boleto-pix.zip).

3. Extraia o conteúdo do .zip baixado e faça upload das pastas `includes` e `modules` para a raíz da sua instalação do WHMCS.

4. Dentro da área administrativa do seu WHMCS: `no menu superior, vá com o mouse sobre a chave de fenda > clique em Apps & Integrations > Pesquise por "PagHiper"`
Após alguns segundos, as opções para `PagHiper Boleto` e `PagHiper PIX` irão aparecer. Clique sobre elas para ativá-las.

### Configuração

5. Para configurar, `passe o mouse sobre a chave de fenda > Clique em "Opções" > Pesquise por "gateways" > Procure pelo módulo PagHiper` que você ativou e clique sobre ele para configurá-lo.

6. **Preencha os campos essenciais para o funcionamento do gateway**: email, API key, token e ID do custom field contendo CPF/CNPJ.

## Solicitação de suporte e novas funcionalidades

Faça a solicitação [clicando aqui](https://github.com/LinkNacional/whmcs-boleto-pix/issues/new) para abrir uma nova issue.

Adicione um título, resumindo o problema e descreve o problema, se possível detalhando com erros e com prints.

# Licença

Copyright 2016-2019 Serviços Online BR.

Licensed under the 3-Clause BSD License (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

[https://opensource.org/licenses/BSD-3-Clause](https://opensource.org/licenses/BSD-3-Clause)

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
## 2.5.3 - 07/06/23

`Bugs resolvidos`

- Corrigir bug ao cancelar fatura quando gateway "paghiper" está desativado

## 2.5.2 - 29/03/23

`Bugs resolvidos`

- Adicionar verificação para razão social
- Adicionar verificação de acesso direto em validate_paghiper_taxid.php

## 2.5.1 - 20/03/23
- Remover código de log
- Adicionar whmcs.json
- Aprimorar documentação no README

## 2.5.0 - 07/03/23

`Novas funcionalidades`

* Configuração para definir valor mínimo e máximo que fatura precisa ter para poder ser paga via Boleto e Pix.
* Configuração para isentar a fatura de taxas, quando o valor da fatura passar valor mínimo.

`Bugs resolvidos`
* Arquivo attach_pdf_slip.php gerando erros em lugares do WHMCS.

## 2.4.2 - 10/01/23

`Bugs resolvidos`

* Label de transação (PIX/boleto) aparecendo trocado em algumas telas de erro
* Erro 0x004681 aparecendo sem motivo em alguns casos
* Problemas intermitentes com a atualização de status das faturas

`Melhorias`

* Mais dados para debug nos Logs de Portais de Pagamento

## 2.4.1 - 09/01/23

`Bugs resolvidos`

* Emissão falhava ao configurar o gateway com um user admin inválido (função de fallback falhava)
* Mensagem "Table exists/Table does not exist" aparecendo de maneira intermitente

## 2.4 - 09/01/23

`Bugs resolvidos`

* Operadores ternários mostravam erros, dependendo da configuração do PHP
* Implementar configuração para desconto por regra

`Melhorias`

* Compatibilidade com PHP 8.1
* Queries convertidas para uso da classe Capsule (ao invés de mysqli)


## 2.3 - 20/06/22

`Bugs resolvidos`

* Transações sendo pagas de maneira duplicada
* Transações PIX sendo exibidas a clientes em alguns casos, ao invés de re-emitidas
* Melhor validação de TaxID (CPF/CNPJ)
* Transações pagáveis sendo ignoradas, caso multa por atraso seja aplicada pelo WHMCS
* Warnings e Deprecated errors
* Dados de clientes sem cadastro não estavam disponíveis para emissão, em algumas circunstâncias
* Suporte a campo de razão social separado
* Número de telefone não estava sendo anexado a transação

`Melhorias`

* Validação de dados e campos aplicada no checkout do front-end
* Mais pontos e informações de log
* Refatoração e melhor eficiência de código


## 2.2.1 - 17/06/21

`Bugs resolvidos`

* Código PIX com quebra de linha indesejada no PDF anexo
* Crash ao marcar pedido como pago ou cancelado via painel
* Conciliação (adição de taxa ou desconto a fatura) impedindo a baixa

## 2.2 - 27/05/21

`Melhorias implementadas`

* Checagem e validação de CPF/CNPJ no checkout, na página de invoice e na tela de boleto/PIX
* Exibição das tags de boleto/PIX no editor de templates de e-mail
* Suporte a campo de razão social (opcional)
* Melhor lógica de reaproveitamento de boletos
* Mais informações nos logs
* Melhor manipulação de CPF/CNPJ para criação de faturas
* Tela de erro, caso o o valor com desconto por pagto. antecipado seja menor que R$ 3
* Tela de erro genérica (evita tela branca, caso uma transação não possa ser gerada)

`Bugs resolvidos`

* Boletos vencidos eram ignorados, ainda que dentro do período de tolerância
* Erro ao cancelar boletos nos logs
* Bloco de inserção de boleto/PIX PDF era executado apenas na primeira fatura da CRON (mod_lsapi)
* Melhor cálculo de desconto para pagto. antecipado
* Maior dinstinção entre as mensagens (evita confusão no front-end)
* Melhor convenção de naming de funções (evita conflitos com outros módulos/gateways)
* Warning de operador ternário removido


## 2.1 - 18/12/20

`Melhorias implementadas`

* Suporte nativo ao PIX PagHiper
* Processo de instalação simplificado
* Novas telas de status e ícones
* Pedidos com boleto agora levam o código de barras junto com a linha digitável
* Melhorias de segurança
* Refatoração completa do plug-in

`Bugs resolvidos`

* Em alguns casos, o campo de nome era utilizado na emissão do boleto, mesmo com CNPJ do cliente informado
* Só exibimos boleto e PIX para usuários com moeda em

## 2.0.3 - 03/01/20

`Melhorias implementadas`

* Cancelamento automático de boletos (na baixa e pagto. parcial de faturas)

`Bugs resolvidos`

* PDF de fatura mostrava boleto em branco de maneira intermitente
* Pagamentos duplicados em alguns ambientes

## 2.0.2 - 14/11/19

`Melhorias implementadas`

* BUGFIX: Faturas parcialmente pagas agora são interpretadas corretamente
* Segurança: Maior proteção contra XSS e SQL injection
* Melhor precisão na determinação de nova data de vencimento, ao reemitir uma fatura vencida
* Novo texto descritivo para a multa proporcional no back-end
* Refatoração geral, redução do número de queries e mais

## 2.0.1.3 - 14/10/19

`Bugs resolvidos`

* Conciliação de multa/desconto impedia baixa das faturas como esperado no WHMCS v7.8

## 2.0.1.2 - 20/05/19

`Bugs resolvidos`

* Boleto não era resgatado caso taxa personalizada estivesse sendo aplicada
* Multa fatorada e descrição fixa não funcionavam em circunstâncias específicas

## 2.0.1.1 - 09/02/19

`Bugs resolvidos`

* Boletos sendo emitidos para todas as faturas, independente do método de pagamento
* Possível erro 500 causado pelo invoicepdf.tpl (Adicionada tag de fechamento do php)

## 2.0.1 - 08/02/19

`Melhorias`

* Agora é possível configurar a nova data de vencimento para reemissão de boletos vencidos
* Possibilidade de restrição na emissão de novos boletos ou emitir para todas as faturas por padrão
* Novas telas de status
* Novo hook para criação do boleto junto com a criação da fatura
* Boleto anexo no e-mail da fatura (veja como ativar aqui)
* Seleção automática de usuário para uso na API Local do WHMCS (necessário para algumas operações internas)
* Novos pontos de log
* Seleção automática de usuário admin para uso na localAPI

`Bugs`

* Melhorada a lógica de busca de boletos emitidos (para faturas vencidas)
* Novo método de formação de URL de retorno

## 2.0 - 01/02/19

`Novas funcionalidades`

* Uso da nova API 2.0
* Re-utilização de boletos
* Juros/multas automáticos na emissão
* Desconto por pagto. antecipadotbladmin

src/modules/gateways/paghiper.php
<?php

// Opções padrão do Gateway
function paghiper_config($params = null)
{
    $custom_fields_conf = paghiper_get_customfield_id();

    $config = [
        'FriendlyName' => [
            'Type' => 'System',
            'Value' => 'PagHiper Boleto',
        ],
        'nota' => [
            'FriendlyName' => 'Nota',
            'Description' => "
            <table>
                <tbody>
                    <tr>
                        <td>Versão <h2 style='font-weight: bold; margin-top: 0px; font-size: 300%;'>2.5.3</h2></td>
                    </tr>
                </tbody>
            </table>

           <h2>Para que o modulo funcione, siga as etapas abaixo:</h2>
           <ul>
               <li>Caso não possua uma conta PagHiper, <a href='https://www.paghiper.com/abra-sua-conta/' target='_blank'><strong> crie a sua aqui</strong></a> <br>
                   Precisa de ajuda para criar sua conta? <a href='https://www.paghiper.com/duvidas/como-se-cadastrar-no-paghiper/' target='_blank'><strong> clique aqui e veja como criar de maneira rápida e facil.</strong></a><br></li>
               <li>Certifique-se que a conta esteja verificada e valida na página de <a href='https://www.paghiper.com/painel/detalhes-da-conta/' target='_blank'><strong>Detalhes da sua conta</strong></a> PagHiper</li>
               <li>Gere o seu token PagHiper na página <a href='https://www.paghiper.com/painel/token/' target='_blank'><strong> Ferramentas > Token</strong></a> e pegue sua ApiKey na página <a href='https://www.paghiper.com/painel/token/' target='_blank'><strong>Minha Conta > Dados da Conta</strong></a></li>
               <li>Ative a integração entre o PagHiper e o <a href='https://www.paghiper.com/painel/whmcs' target='_blank'><strong>WHMCS</strong></a>, <a href='https://www.paghiper.com/painel/whmcs' target='_blank'><strong>Acesse aqui</strong></a> e ative.</li>
               <li><h5>Suporte</h5><p>Se tiver qualquer duvida, visite a nossa <a href='https://www.paghiper.com/atendimento/' target='_blank'><strong>central de atendimento</strong></a></p></li>
           </ul>",
        ],

        'email' => [
            'FriendlyName' => 'Email',
            'Type' => 'text',
            'Size' => '100',
            'Description' => 'Email da conta PagHiper que irá receber',
        ],
        'api_key' => [
            'FriendlyName' => 'API Key',
            'Type' => 'text',
            'Size' => '66',
            'Description' => "Campo composto de números, letras, traços e hífen.
Sempre começa por apk_. Caso não tenha essa informação, pegue sua chave API <a href='https://www.paghiper.com/painel/credenciais/' target='_blank'><strong>aqui</strong></a>.",
        ],
        'token' => [
            'FriendlyName' => 'Token',
            'Type' => 'text',
            'Size' => '66',
            'Description' => "Extremamente importante, você pode gerar seu token em nossa pagina: Painel > Ferramentas > Token ( <a href='https://www.paghiper.com/painel/token/' target='_blank'><strong>Confira Aqui</strong></a> ).",
        ],
        'cpf_cnpj' => [
            'FriendlyName' => 'ID do custom field contendo CPF/CNPJ',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Defina aqui o ID do campo usado para coletar CPF/CNPJ do seu cliente. Isso é necessário para usar o checkout transparente.' . $custom_fields_conf,
        ],
        'razao_social' => [
            'FriendlyName' => 'ID do custom field contendo Razão Social',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Defina aqui o ID do campo usado, caso utilize um campo personalizado para coletar a Razão Social do seu cliente. Isso é opcional.',
        ],
        'porcento' => [
            'FriendlyName' => 'Taxa Percentual (%)',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Porcentagem da fatura a se pagar a mais por usar o PagHiper. Ex.: (2.5). Obs.: não precisa colocar o % no final. Obs²: Use o ponto (.) como delimitador de casas decimais. <br> Recomendamos não cobrar nenhuma taxa.',
        ],
        'taxa' => [
            'FriendlyName' => 'Taxa fixa',
            'Type' => 'text',
            'Size' => '7',
            'Description' => 'Taxa cobrada a mais do cliente por utilizar esse meio de pagamento, exemplo: 2.0 (dois reais). Obs: Use o ponto (.) como delimitador de casas decimais.<br> Recomendamos não cobrar nenhuma taxa.',
        ],
        'abrirauto' => [
            'FriendlyName' => 'Abrir boleto ao abrir fatura?',
            'Type' => 'yesno',
        ],
        'fixed_description' => [
            'FriendlyName' => 'Exibe ou não a frase fixa do boleto (configurada no painel da PagHiper)',
            'Type' => 'yesno',
        ],
        'open_after_day_due' => [
            'FriendlyName' => 'Tolerância para pagto',
            'Type' => 'text',
            'Size' => '2',
            'Description' => 'Número máximo de dias em que o boleto poderá ser pago após o vencimento. (Prática comum para quem opta por cobrar juros e multas).',
        ],
        'reissue_unpaid' => [
            'FriendlyName' => 'Vencimento padrão para boletos emitidos',
            'Type' => 'dropdown',
            'Options' => [
                '-1' => 'Não permitir reemissão',
                '0' => 'Vcto. no mesmo dia',
                '1' => '+1 dia',
                '2' => '+2 dias',
                '3' => '+3 dias',
                '4' => '+4 dias',
                '5' => '+5 dias',
            ],
            'Description' => 'Escolha a quantidade de dias para o vencimento de boletos reemitidos (para faturas ja vencidas). Caso decida não permitir reemissão, você precisará mudar a data de vencimento manualmente.',
        ],
        'late_payment_fine' => [
            'FriendlyName' => 'Percentual da multa por atraso (%)',
            'Type' => 'text',
            'Size' => '1',
            'Description' => 'O percentual máximo autorizado é de 2%, de acordo artigo 52, parágrafo primeiro do Código de Defesa do Consumidor, Lei 8.078/90',
        ],
        'per_day_interest' => [
            'FriendlyName' => 'Juros proporcional',
            'Type' => 'yesno',
            'Description' => 'Ao aplicar 1% de juros máximo ao mês, esse percentual será cobrado proporcionalmente aos dias de atraso.<br><br>Dividindo 1% por 30 dias = 0,033% por dia de atraso.',
        ],
        'early_payment_discounts_days' => [
            'FriendlyName' => 'Qtde. de dias para aplicação de desconto',
            'Type' => 'text',
            'Size' => '2',
            'Description' => 'Número de dias em que o pagamento pode ser realizado com antecedência recebendo o desconto extra.',
        ],
        'early_payment_discounts_cents' => [
            'FriendlyName' => 'Desconto por pagto. antecipado',
            'Type' => 'text',
            'Size' => '6',
            'Description' => 'Valor do desconto que será aplicado caso o pagamento ocorra de forma antecipada. Em percentual (Ex.: 10%)',
        ],
        'issue_all' => [
            'FriendlyName' => 'Gerar boletos para todos os pedidos?',
            'Type' => 'dropdown',
            'Options' => [
                '1' => 'Sim',
                '0' => 'Não',
            ],
            'Description' => 'Caso selecione não, boletos bancários e lihnas digitáveis serão selecionadas somente caso o cliente selecione "Boleto Bancário" (ou o nome que você configurar no primeiro campo de configuração) como método de pagamento padrão.',
        ],
        'tax_id_validation' => [
            'FriendlyName' => 'Validar campos de CPF/CNPJ no checkout?',
            'Type' => 'dropdown',
            'Options' => [
                'strict' => 'Aceitar pedidos somente com CPF e CNPJ válidos',
                'flexible_f' => 'Aceitar pedidos, desde que CPF seja válido',
                'flexible_j' => 'Aceitar pedidos, desde que CNPJ seja válido',
                'flexible' => 'Aceitar pedidos com CPF ou CNPJ válidos, tanto faz',
                '0' => 'Não validar nenhum dos campos',
            ],
            'Description' => 'Caso selecione não, boletos bancários e lihnas digitáveis serão selecionadas somente caso o cliente selecione "Boleto Bancário" (ou o nome que você configurar no primeiro campo de configuração) como método de pagamento padrão.',
        ],
        'maximum_payment_amount' => [
            'FriendlyName' => 'Valor máximo para pagamento',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Ex: 50,00 e 5000,00 - utilizar apenas números e vírgulas.',
        ],
        'minimum_payment_amount' => [
            'FriendlyName' => 'Valor mínimo para pagamento',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Ex: 5,00 e 50,00 - utilizar apenas números e vírgulas.',
        ],
        'min_amount_for_taxes_exempt' => [
            'FriendlyName' => 'Isentar de taxas',
            'Type' => 'text',
            'Size' => '3',
            'Description' => 'Não aplicar taxas (Taxa percentual + Taxa fixa) quando o valor da fatura for maior ou igual a este. Ex: 100,00 (usar apenas vírgulas).',
        ],
        'admin' => [
            'FriendlyName' => 'Administrador atribuído',
            'Type' => 'text',
            'Size' => '10',
            'Default' => 'admin',
            'Description' => 'Insira o nome de usuário ou ID do administrador do WHMCS que será atribuído as transações. Necessário para usar a API interna do WHMCS.'
        ],
        'suporte' => [
            'FriendlyName' => "<span class='label label-primary'><i class='fa fa-question-circle'></i> Suporte</span>",
            'Description' => '<h2>Para informações ou duvidas: </h2><br><br>
<ul>
<li>Duvidas sobre a conta <strong> PAGHIPER:</strong> <br><br>
Devem ser resolvidas diretamente na central de atendimento: <br>
<strong><a href="https://www.paghiper.com/atendimento" target="_blank">https://www.paghiper.com/atendimento</a></strong></li>
<br><br><br>
<li>Duvidas sobre o <strong> Modulo WHMCS </strong> <br><br>
Tem uma dúvida ou quer contribuir para o projeto? Acesse nosso repositório no GitHub!
<br>
<strong><a href="https://github.com/paghiper/whmcs" target="_blank">https://github.com/paghiper/whmcs</a></strong></li>
</ul><br>',
        ],

    ];

    return $config;
}

function paghiper_link($params)
{
    $isElegibleForPayment = paghiper_is_balance_elegible_for_payment('paghiper', $params['invoiceid']);

    if (is_string($isElegibleForPayment)) {
        return $isElegibleForPayment;
    }

    // Definimos os dados para retorno e checkout.
    $systemurl = rtrim($params['systemurl'], '/');
    $urlRetorno = $systemurl . '/modules/gateways/' . basename(__FILE__);
    $customFields = explode('|', $params['cpf_cnpj']); // id dos campos de acordo com configuração do whmcs

    // Abrir o boleto automaticamente ao abrir a fatura
    if($params['abrirauto'] == true):
        $target = '';
        $abrirAuto = "<script type='text/javascript'> document.paghiper.submit()</script>";
    else:
        $target = "target='_blank'";
        $abrirAuto = '';
    endif;

    // Checamos o CPF/CNPJ novamente, para evitar problemas no checkout
    $taxIdFields = explode('|', $params['cpf_cnpj']);
    $payerNameField = $params['razao_social'];

    $clientCustomFields = [];
    foreach ($params['clientdetails']['customfields'] as $key => $value) {
        $clientCustomFields[$value['id']] = $value['value'];
    }

    $clientTaxIds = [];
    if (count($taxIdFields) > 1) {
        $clientTaxIds[] = $clientCustomFields[$taxIdFields[0]];
        $clientTaxIds[] = $clientCustomFields[$taxIdFields[1]];
    } else {
        $clientTaxIds[] = $clientCustomFields[$taxIdFields[0]];
    }

    $isValidTaxId = false;
    foreach ($clientTaxIds as $clientTaxId) {
        if (paghiper_is_tax_id_valid($clientTaxId)) {
            $isValidTaxId = true;

            break ;
        }
    }

    $code = '';

    $isValidPayerName = true;
    $clientPayerName = $clientCustomFields[$payerNameField];
    foreach($clientTaxIds as $clientTaxId) {
        $taxid_value = preg_replace('/\D/', '', $clientTaxId);

        if(strlen($taxid_value) > 11 && empty($params['clientdetails']['companyname']) && empty($payerNameField) && empty($clientPayerName)) {
            $isValidPayerName = false;
            $code .= sprintf('<div class="alert alert-danger" role="alert">%s</div>', 'Razão social inválida, atualize seus dados cadastrais.');
        }
    }

    if($isValidTaxId) {
        $client_details = [
            'firstname' => $params['clientdetails']['firstname'],
            'lastname' => $params['clientdetails']['lastname'],
            'companyname' => $params['clientdetails']['companyname'],
            'email' => $params['clientdetails']['email'],
            'phonenumber' => $params['clientdetails']['phonenumber'],
            'address1' => $params['clientdetails']['address1'],
            'address2' => $params['clientdetails']['address2'],
            'city' => $params['clientdetails']['city'],
            'state' => $params['clientdetails']['state'],
            'postcode' => $params['clientdetails']['postcode'],
            'cpf_cnpj' => $clientTaxId,
            'razao_social' => $clientPayerName
        ];

        if($isValidPayerName) {
            // Código do checkout
            $code .= "<!-- INICIO DO FORM DO BOLETO PAGHIPER -->
            <form name=\"paghiper\" action=\"{$urlRetorno}?invoiceid={$params['invoiceid']}&uuid={$params['clientdetails']['userid']}&mail={$params['clientdetails']['email']}\" method=\"post\">
                <input type=\"hidden\" name=\"client_data\" value='" . json_encode($client_details) . "'>
                <input type='image' src='{$systemurl}/modules/gateways/paghiper/assets/img/billet.jpg' title='Pagar com Boleto' alt='Pagar com Boleto' border='0' align='absbottom' width='120' height='74' /><br>
                <button formtarget='_blank' class='btn btn-success' style='margin-top: 5px;' type=\"submit\"><i class='fa fa-barcode'></i> Gerar Boleto</button>
                <br> <br>
                <div class='alert alert-warning' role='alert'>
                <strong>Importante:</strong> A compensação bancária poderá levar até 2 dias úteis.
                </div>
                <!-- FIM DO BOLETO PAGHIPER -->
            </form>
            {$abrirAuto}";
        } else {
            $code = sprintf('<div class="alert alert-danger" role="alert">%s</div>', 'CPF ou CNPJ inválido, atualize seus dados cadastrais.');
        }
    } else {
        $code .= sprintf('<div class="alert alert-danger" role="alert">%s</div>', 'CPF ou CNPJ inválido, atualize seus dados cadastrais.');
    }

    return $code;
}

$is_pix = false;

require_once __DIR__ . '/paghiper/inc/helpers/gateway_functions.php';
require_once __DIR__ . '/paghiper/inc/helpers/process_payment.php';
